File: 0413-Arithmetic-Slices/lc0413.cc
// C++ solution.

// DP.
// f[i]: the number of arithmetic slices ending with A[i].
// if A[i-2], A[i-1], A[i] are arithmetic, f[i] = f[i-1] + 1.
class Solution {
public:
    int numberOfArithmeticSlices(vector<int>& A) {
        int len = A.size();
        if (len < 3) { return 0; }
        vector<int> dp(len, 0);
        int ans = 0;
        for (int i = 2; i < len; i++) {
            if (A[i]-A[i-1] == A[i-1]-A[i-2]) {
                dp[i] = dp[i-1] + 1;
            }
            ans += dp[i];
        }
        return ans;
    }
};

File: 0413-Arithmetic-Slices/lc0413.go
// Go solution.

// DP method
func numberOfArithmeticSlices(nums []int) int {
    numsLen := len(nums)
    if numsLen < 3 {
        return 0
    }

    dp := make([]int, numsLen)
    ans := 0
    for i := 2; i < numsLen; i++ {
        if nums[i]-nums[i-1] == nums[i-1]-nums[i-2] {
            dp[i] = dp[i-1] + 1
        }
        ans += dp[i]
    }
    return ans
}

// DP method with O(1) space complexity
func numberOfArithmeticSlices(nums []int) int {
    numsLen := len(nums)
    if numsLen < 3 {
        return 0
    }

    cnt := 0
    ans := 0
    for i := 2; i < numsLen; i++ {
        if nums[i]-nums[i-1] == nums[i-1]-nums[i-2] {
            cnt++
        } else {
            cnt = 0
        }
        ans += cnt
    }
    return ans
}